res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1136 @@
+            var outval = this.outValueScale(value);
+            // NOTE: outputColor only handles hotcueColors
+            // and there is no hotcueColor for turning the LED
+            // off. So the `send()` function is responsible for turning the
+            // actual LED off.
+            if (this.colorKey !== undefined && outval !== this.off) {
+                this.outputColor(engine.getValue(this.group, this.colorKey));
+            } else {
+                this.send(outval);
+            }
+        }
+        outputColor(colorCode) {
+            // Sends the color from the colorCode to the controller. This
+            // method will not be called if no colorKey has been specified.
+            if (colorCode === undefined || colorCode < 0 || colorCode > 0xFFFFFF) {
+                print("Ignoring invalid color code '" + colorCode + "' in outputColor()");
+                return;
+            }
+
+            if (this.colorMapper !== undefined) {
+                // This HotcueButton holds a reference to a ColorMapper. This means
+                // that the controller only supports a fixed set of colors, so we
+                // get the MIDI value for the nearest supported color and send it.
+                var nearestColorValue = this.colorMapper.getValueForNearestColor(colorCode);
+                this.send(nearestColorValue);
+            } else {
+                // Since outputColor has been called but no ColorMapper is
+                // available, we can assume that controller supports arbitrary
+                // RGB color output.
+                this.sendRGB(colorCodeToObject(colorCode));
+            }
+        }
+        sendRGB(_colorObject) {
+            // This method needs to be overridden in controller mappings,
+            // because the procedure is controller-dependent.
+            throw "sendRGB(colorObject) not implemented -- unable to send RGB colors";
+        }
+        connect() {
+            super.connect();
+            if (undefined !== this.group && this.colorKey !== undefined) {
+                this.connections[1] = engine.makeConnection(this.group, this.colorKey, function(color) {
+                    if (engine.getValue(this.group, this.outKey)) {
+                        this.outputColor(color);
+                    }
+                }.bind(this));
+            }
+        }
+    };
+
+    var SamplerButton = class extends Button {
+        constructor(options) {
+            if (options.number === undefined) {
+                throw "No sampler number specified for SamplerButton constructor";
+            }
+            options.group = "[Sampler" + options.number + "]";
+            options.outKey = null, // hack to get Component constructor to call connect()
+            super(options);
+        }
+        unshift() {
+            this.input = function(channel, control, value, status, _group) {
+                if (this.isPress(channel, control, value, status)) {
```

Comment:
This method name has bugged me for a while. Is it a weird way to abbreviate "isPressed"? Is it used to detect if a value is a "press" event instead of a release? Can we find another name?

Reply:
I don't think its supposed to be an abbreviation, the `isPress` function is supposed to determine whether a certain message conveys a "press of the button". Would `isPressed` a better name?
I guess something like `isButtonDown` would be more standard?